Army-Java/Ability/Ability.java

```diff
@@ -0,0 +1,65 @@
+package com.gymfox.Army.Ability;
+
+import com.gymfox.Army.Exception.IsSelfAttackException;
+import com.gymfox.Army.Exception.UnitIsDeadException;
+import com.gymfox.Army.Spells.Spell;
+import com.gymfox.Army.State.State;
+import com.gymfox.Army.Units.Unit;
+
+public abstract class Ability {
+    private Unit currentUnit;
+
+    public Ability(Unit currentUnit) {
+        this.currentUnit = currentUnit;
+    }
+
+    public void attack(Unit victim) throws UnitIsDeadException, IsSelfAttackException {
+        victim.takeDamage(currentUnit.getDamage());
+        victim.counterAttack(currentUnit);
+    }
+
+    public void counterAttack(Unit victim) throws UnitIsDeadException {
+        victim.takeDamage(currentUnit.getDamage() / 2);
+
+        if ( currentUnit.getUnitType() == Unit.UnitType.VAMPIRE ) {
+            int recoverHP = victim.getCurrentHP() / 10;
+
+            currentUnit.heal(recoverHP);
+            victim.takeDamage(recoverHP);
+        }
+    }
+
+    public void takeDamage(int damage) throws UnitIsDeadException {
+        currentUnit.getCurrentState().takeDamage(damage);
+    }
+
+    public void takeMagicDamage(int damage) throws UnitIsDeadException {
+        currentUnit.getCurrentState().takeMagicDamage(damage);
```

Comment:
So as I already told, as for me it makes sense to keep states in abilites, not in units